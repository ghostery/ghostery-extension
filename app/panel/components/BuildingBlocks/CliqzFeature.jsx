/**
 * Cliqz Features Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2019 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import ClassNames from 'classnames';
import Tooltip from '../Tooltip';

/**
 * @class Implements rendering and interaction for Cliqz feature icon toggles
 * @memberof PanelBuildingBlocks
 */
class CliqzFeature extends React.Component {
	constructor(props) {
		super(props);

		this.clickCliqzFeature = this.clickCliqzFeature.bind(this);
	}

	/**
	 * Handles clicks on the Cliqz feature icon, toggling it on/off
	 */
	clickCliqzFeature() {
		const {
			active,
			clickButton,
			cliqzInactive,
			type
		} = this.props;

		if (cliqzInactive) {
			return;
		}

		const featureType = type === 'anti_track' ? 'anti_tracking' : type;

		clickButton({
			feature: `enable_${featureType}`,
			status: active,
			text: this._getAlertText(active, type),
		});
	}

	_getStatus(active) {
		return active ? t('on') : t('off');
	}

	_getTooltipBodyText(active, isTooltipBody, type) {
		if (!isTooltipBody) return false;

		if (active) {
			switch (type) {
				case 'ad_block':
					return t('tooltip_ad_block_on');
				case 'anti_track':
					return t('tooltip_anti_track_on');
				case 'smart_block':
					return t('tooltip_smart_block_on');
				default:
					return false;
			}
		} else {
			switch (type) {
				case 'ad_block':
					return t('ad_blocking_DESC');
				case 'anti_track':
					return t('anti_tracking_DESC');
				case 'smart_block':
					return t('smart_blocking_DESC');
				default:
					return false;
			}
		}
	}

	_getTooltipHeaderText(isTooltipHeader, type) {
		if (!isTooltipHeader) return false;

		switch (type) {
			case 'ad_block':
				return t('enhanced_ad_blocking');
			case 'anti_track':
				return t('enhanced_anti_tracking');
			case 'smart_block':
				return t('smart_blocking');
			default:
				return false;
		}
	}

	_getAlertText(active, type) {
		return active ?
			t(`alert_${type}_off`) :
			t(`alert_${type}_on`);
	}

	/**
	 * React's required render function. Returns JSX
	 * @return {JSX} JSX for rendering a Cliqz Feature icon toggle
	 */
	render() {
		const {
			active,
			cliqzInactive,
			isSmaller,
			isCondensed,
			isTiny,
			isTooltipBody,
			isTooltipHeader,
			tooltipPosition,
			type,
		} = this.props;

		const cliqzFeatureClassNames = ClassNames('CliqzFeature', {
			'CliqzFeature--normal': !isSmaller && !isCondensed,
			'CliqzFeature--smaller': isSmaller,
			'CliqzFeature--condensed': isCondensed,
			'CliqzFeature--tiny': isTiny,
			'CliqzFeature--active': active,
			'CliqzFeature--inactive': !active,
			clickable: !cliqzInactive,
			'not-clickable': cliqzInactive,
		});
		const cssTypeName = `CliqzFeature__icon--${type.replace('_', '-')}`;
		const iconClassNames = ClassNames('CliqzFeature__icon', cssTypeName, 'g-tooltip');

		const featureType = type === 'anti_track' ? 'anti_tracking' : type;
		let featureName;
		if (featureType === 'anti_tracking') {
			featureName = t('enhanced_anti_tracking');
		} else if (featureType === 'ad_block') {
			featureName = t('enhanced_ad_blocking');
		} else if (featureType === 'smart_block') {
			featureName = t('smart_blocking');
		}

		return (
			<div className={cliqzFeatureClassNames} onClick={this.clickCliqzFeature}>
				<div className="CliqzFeature__status">{this._getStatus(active)}</div>
				<div className={iconClassNames}>
					<Tooltip
						header={this._getTooltipHeaderText(isTooltipHeader, type)}
						body={this._getTooltipBodyText(active, isTooltipBody, type)}
						position={tooltipPosition}
						// className={isTiny ? 'CliqzFeature--tooltipUp' : ''}
					/>
				</div>
				<div className="CliqzFeature__feature-name">
					{featureName}
				</div>
			</div>
		);
	}
}

export default CliqzFeature;
