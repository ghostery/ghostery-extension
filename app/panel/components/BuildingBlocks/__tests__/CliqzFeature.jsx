/**
 * Cliqz Feature Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import CliqzFeature from '../CliqzFeature';

// Fake the translation function to only return the translation key
global.t = function(str) {
	return str;
};

// Fake the Tooltip implementation
jest.mock('../../../../shared-components/Tooltip');

describe('app/panel/components/CliqzFeature.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('CliqzFeature is rendered correctly with falsy props', () => {
			const component = renderer.create(
				<div>
					<CliqzFeature
						clickButton={() => {}}
						type="anti_track"
						active={false}
						cliqzInactive={false}
						isSmaller={false}
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="ad_block"
						active={false}
						cliqzInactive={false}
						isSmaller={false}
						isCondensed={false}
						isTooltipHeader={false}
						isTooltipBody={false}
						tooltipPosition=""
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="smart_block"
						active={false}
						cliqzInactive={false}
						isSmaller={false}
						isCondensed={false}
						isTooltipHeader={false}
						isTooltipBody={false}
						tooltipPosition=""
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('CliqzFeature is rendered correctly with some truthy props', () => {
			const component = renderer.create(
				<div>
					<CliqzFeature
						clickButton={() => {}}
						type="anti_track"
						active
						cliqzInactive={false}
						isSmaller={false}
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="ad_block"
						active
						cliqzInactive
						isSmaller={false}
						isCondensed={false}
						isTooltipHeader={false}
						isTooltipBody={false}
						tooltipPosition=""
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="smart_block"
						active
						cliqzInactive
						isSmaller={false}
						isCondensed
						isTooltipHeader
						isTooltipBody
						tooltipPosition="top"
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('CliqzFeature is rendered correctly with all truthy props', () => {
			const component = renderer.create(
				<div>
					<CliqzFeature
						clickButton={() => {}}
						type="anti_track"
						active
						cliqzInactive
						isSmaller
						isCondensed
						isTooltipHeader
						isTooltipBody
						tooltipPosition="right"
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="ad_block"
						active
						cliqzInactive
						isSmaller
						isCondensed
						isTooltipHeader
						isTooltipBody
						tooltipPosition="top"
					/>
					<CliqzFeature
						clickButton={() => {}}
						type="smart_block"
						active
						cliqzInactive
						isSmaller
						isCondensed
						isTooltipHeader
						isTooltipBody
						tooltipPosition="top"
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('CliqzFeature handles clicks correctly', () => {
			const clickButton = jest.fn();
			const component = shallow(
				<CliqzFeature
					clickButton={clickButton}
					type="smart_block"
					active
					cliqzInactive
					isSmaller
				/>
			);
			expect(clickButton.mock.calls.length).toBe(0);
			component.find('.CliqzFeature').simulate('click');
			expect(clickButton.mock.calls.length).toBe(0);

			component.setProps({ cliqzInactive: false });
			component.find('.CliqzFeature').simulate('click');

			component.setProps({ active: false });
			component.find('.CliqzFeature').simulate('click');
			expect(clickButton.mock.calls.length).toBe(2);
			expect(clickButton.mock.calls[0][0].status).toBe(true);
			expect(clickButton.mock.calls[1][0].status).toBe(false);
		});
	});
});
