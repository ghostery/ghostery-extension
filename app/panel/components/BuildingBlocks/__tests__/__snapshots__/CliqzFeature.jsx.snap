// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel/components/CliqzFeature.jsx Snapshot tests with react-test-renderer CliqzFeature is rendered correctly with all truthy props 1`] = `
<div>
  <div
    className="CliqzFeature CliqzFeature--smaller CliqzFeature--condensed CliqzFeature--active not-clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--anti-track g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_anti_tracking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--smaller CliqzFeature--condensed CliqzFeature--active not-clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--ad-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_ad_blocking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--smaller CliqzFeature--condensed CliqzFeature--active not-clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--smart-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      smart_blocking
    </div>
  </div>
</div>
`;

exports[`app/panel/components/CliqzFeature.jsx Snapshot tests with react-test-renderer CliqzFeature is rendered correctly with falsy props 1`] = `
<div>
  <div
    className="CliqzFeature CliqzFeature--normal CliqzFeature--inactive clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      off
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--anti-track g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_anti_tracking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--normal CliqzFeature--inactive clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      off
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--ad-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_ad_blocking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--normal CliqzFeature--inactive clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      off
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--smart-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      smart_blocking
    </div>
  </div>
</div>
`;

exports[`app/panel/components/CliqzFeature.jsx Snapshot tests with react-test-renderer CliqzFeature is rendered correctly with some truthy props 1`] = `
<div>
  <div
    className="CliqzFeature CliqzFeature--normal CliqzFeature--active clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--anti-track g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_anti_tracking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--normal CliqzFeature--active not-clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--ad-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      enhanced_ad_blocking
    </div>
  </div>
  <div
    className="CliqzFeature CliqzFeature--condensed CliqzFeature--active not-clickable"
    onClick={[Function]}
  >
    <div
      className="CliqzFeature__status"
    >
      on
    </div>
    <div
      className="CliqzFeature__icon CliqzFeature__icon--smart-block g-tooltip"
    />
    <div
      className="CliqzFeature__feature-name"
    >
      smart_blocking
    </div>
  </div>
</div>
`;
